### DISCUSSION TITLE
Plugin error no such file or directory example.app

### DISCUSSION THREAD

**Discussion by alexa22013** [emqx/emqx#11233] [Q&A] [answered]
I am building a plugin using the emqx-plugin-template. For any dependency I add I get the error:

`2023-07-08T22:14:59.116350+00:00 alert: {"time":1688854499116350,"reason":{"reason":"{\"no such file or directory\",\"sample_dependency.app\"}","error":"failed_to_start_plugin_app","err_app":"sample_dependency","app":"my_plugin"},"pid":"<0.2219.0>","msg":"failed_to_start_plugin","line":356,"level":"alert","mfa":"emqx_plugins:ensure_started/1"}`

Need help to solve this issue..

Thanks in advance

**Comment by SergeTupchiy** [comment]
Hi @alexa22013 ,
I suspect that the reason was that the app wasn't listed in `relx` config in `rebar.config`.
For example:
```
{relx, [ {release, {my_emqx_plugin, "1.0.0"}, %% this is the release version, different from app vsn in .app file
            [ my_emqx_plugin
            , map_sets
            , <your dependency application name>
            ]}
       , {dev_mode, false}
       , {include_erts, false}
       ]}.
```

**Reply by alexa22013** [reply]
Thanks @SergeTupchiy for the reply, I have listed in relx config in rebar.config but still I get the same error.

**Comment by SergeTupchiy** [comment]
@alexa22013, would you be able to provide a minimum reproducible example with your plugin and dependency applications?

**Reply by alexa22013** [reply]
@SergeTupchiy this is the rebar.config and plugin.app.src. I am using erlkaf as dependency. I have only modified these files to add the dependency. 

rebar.config
```
%% -*- mode: erlang -*-
{deps,
    [ %% not all plugins have to depend on emqx
      %% but if they do, use git_subdir in the default profile
      {emqx, {git_subdir, "https://github.com/emqx/emqx.git", {tag, "v5.1.0"}, "apps/emqx"}}
    , {emqx_ctl, {git_subdir, "https://github.com/emqx/emqx.git", {tag, "v5.1.0"}, "apps/emqx_ctl"}}
    , {emqx_utils, {git_subdir, "https://github.com/emqx/emqx.git", {tag, "v5.1.0"}, "apps/emqx_utils"}}
      %% this is my plugin's dependency
    , {map_sets, "1.1.0"}
    , {erlkaf, "2.1.4"}
    ]}.

{plugins, [
    {emqx_plugrel, {git, "https://github.com/emqx/emqx_plugrel.git", {tag, "0.3.0"}}}
]}.

{erl_opts, [debug_info]}.

{relx, [ {release, {kafka_bridge, "1.0.0"}, %% this is the release version, different from app vsn in .app file
            [ kafka_bridge
            , map_sets
            , erlkaf
            ]}
       , {dev_mode, false}
       , {include_erts, false}
       ]}.

%% Additional info of the plugin
{emqx_plugrel,
    [ {authors, ["Alexa"]}
    , {builder,
        [ {name, "Alexa"}
        , {contact, "alexa21@gmail.com"}
        , {website, "http://example.com"}
        ]}
    , {repo, "https://github.com/emqx/emqx-plugin-template"}
    , {functionality, ["Bridge"]}
    , {compatibility,
        [ {emqx, "~> 5.0"}
        ]}
    , {description, "Another amazing EMQX plugin."}
    ]
}.

{xref_checks,[undefined_function_calls,undefined_functions,locals_not_used,
              deprecated_function_calls,warnings_as_errors,deprecated_functions]}.

``` 

plugin.app.src
```
{application, kafka_bridge,
 [{description, "Another amazing EMQX plugin."},
  {vsn, "0.1.0"},
  {modules, []},
  {registered, [kafka_bridge_sup]},
  {applications, [kernel, stdlib, erlkaf, map_sets]},
  {mod, {kafka_bridge_app, []}},
  {env, []},
  {licenses, ["Apache-2.0"]},
  {maintainers, ["Alexa <alexa21@gmail.com>"]}
 ]}.

```

**Comment by SergeTupchiy** [comment]
I cannot reproduce `"no such file or directory"` error, maybe try to remove `_build` directory and re-build the plugin from scratch?
However, please note that erlkaf uses NIF which is currently not supported in EMQX plugins, so it's expected to ultimately fail with another error reason, something like: 
```
2023-07-10T13:49:08.956331+03:00 [error] Error in process <0.3641.0> on node 'emqx@127.0.0.1' with exit value:, {undef,[{erlkaf_utils,get_priv_path,[erlkaf_nif],[]},{erlkaf_nif,load_nif,0,[{file,"/tmp/my_emqx_plugin/_build/default/lib/erlkaf/src/erlkaf_nif.erl"},{line,29}]},{code_server,'-handle_on_load/5-fun-0-',1,[{file,"code_server.erl"},{line,1317}]}]}
2023-07-10T13:49:08.961472+03:00 [alert] msg: failed_to_start_plugin, mfa: emqx_plugins:ensure_started/1, line: 361, reason: #{error => "failed_to_load_plugin_beam",path => "plugins/my_emqx_plugin-1.0.0/erlkaf-2.1.4/ebin/erlkaf_nif.beam",reason => on_load_failure}
```
It's a known issue which is planned to be fixed (https://github.com/emqx/emqx/issues/10399).

**Reply by alexa22013** [reply]
Thanks! @SergeTupchiy for making aware of the issue, I also tried using the brod kafka client but still I get hte same issue. I am compiling the plugin on a mac, I believe its not a cross compilation issue because the plugin works fine without any dependency added to the plugin. The file which is said to be not found in the error is actually present in the _build directory. But for some reason its not able to find it.

**Reply by alexa22013** [reply]
@SergeTupchiy I also tried removing _build directory and building from scratch, still same issue.

**Reply by SergeTupchiy** [reply]
@alexa22013 , which EMQX version are you using?

**Reply by alexa22013** [reply]
@SergeTupchiy I am using emqx v5.1.0

**Reply by SergeTupchiy** [reply]
@alexa22013, can you try the following:
- add only `brod` to `src/<your_plugin>.app.src` and `rebar.config` `deps`, e.g.:
```
{application, my_emqx_plugin,
 [{description, "Another amazing EMQX plugin."},
  {vsn, "0.1.0"},
  {modules, []},
  {registered, [my_emqx_plugin_sup]},
  {applications, [kernel, stdlib, map_sets, brod]},
  {mod, {my_emqx_plugin_app, []}},
  {env, []},
  {licenses, ["Apache-2.0"]},
  {maintainers, []}
 ]}.
```
```
{deps,
    [ %% not all plugins have to depend on emqx
      %% but if they do, use git_subdir in the default profile
      {emqx, {git_subdir, "https://github.com/emqx/emqx.git", {tag, "v5.1.0"}, "apps/emqx"}}
    , {emqx_ctl, {git_subdir, "https://github.com/emqx/emqx.git", {tag, "v5.1.0"}, "apps/emqx_ctl"}}
    , {emqx_utils, {git_subdir, "https://github.com/emqx/emqx.git", {tag, "v5.1.0"}, "apps/emqx_utils"}}
      %% this is my plugin's dependency
    , {map_sets, "1.1.0"}
    , {brod, {git, "git@github.com:kafka4beam/brod.git", {tag, "3.16.8"}}}
    ]}.
```
- list brod as well as its own dependencies in `relx` rebar.config:
```
{relx, [ {release, {my_emqx_plugin, "1.0.0"}, %% this is the release version, different from app vsn in .app file
            [ my_emqx_plugin
            , map_sets
            , brod
            , kafka_protocol
            , crc32cer
            , snappyer
            , supervisor3
            ]}
       , {dev_mode, false}
       , {include_erts, false}
       ]}.
```
- build it with `make rel`
- upload and start the plugin in EMQX 
Please share logs (if any).

**Reply by alexa22013** [reply]
Thanks! @SergeTupchiy, Iam able to upload and start the plugin, it was able to resolve brod. But on producing message I get these errors

```
2023-07-11T20:58:58.909219+00:00 error: {"time":1689109138909219,"pid":"<0.30073.21>","msg":"supervisor: {<0.30073.21>,brod_producers_sup}, errorContext: child_terminated, reason: {{crc32cer_nif_not_loaded,\"plugins/kafka_bridge-1.0.0/crc32cer-0.1.8/priv/crc32cer_nif\"},[{erlang,nif_error,[{crc32cer_nif_not_loaded,\"plugins/kafka_bridge-1.0.0/crc32cer-0.1.8/priv/crc32cer_nif\"}],[{error_info,#{module => erl_erts_errors}}]},{crc32cer,nif,2,[{file,\"/Users/alexa/Desktop/kafka_bridge/_build/default/lib/crc32cer/src/crc32cer.erl\"},{line,19}]},{kpro_batch,encode_tx,4,[{file,\"/Users/alexa/Desktop/kafka_bridge/_build/default/lib/kafka_protocol/src/kpro_batch.erl\"},{line,96}]},{kpro_req_lib,produce,5,[{file,\"/Users/alexa/Desktop/kafka_bridge/_build/default/lib/kafka_protocol/src/kpro_req_lib.erl\"},{line,286}]},{brod_producer,do_send_fun,4,[{file,\"/Users/alexa/Desktop/kafka_bridge/_build/default/lib/brod/src/brod_producer.erl\"},{line,428}]},{brod_producer_buffer,do_send,4,[{file,\"/Users/alexa/Desktop/kafka_bridge/_build/default/lib/brod/src/brod_producer_buffer.erl\"},{line,276}]},{brod_producer,maybe_produce,1,[{file,\"/Users/alexa/Desktop/kafka_bridge/_build/default/lib/brod/src/brod_producer.erl\"},{line,536}]},{brod_producer,do_handle_produce,3,[{file,\"/Users/alexa/Desktop/kafka_bridge/_build/default/lib/brod/src/brod_producer.erl\"},{line,488}]},{gen_server,try_dispatch,4,[{file,\"gen_server.erl\"},{line,1123}]},{gen_server,handle_msg,6,[{file,\"gen_server.erl\"},{line,1200}]},{proc_lib,init_p_do_apply,3,[{file,\"proc_lib.erl\"},{line,240}]}]}, offender: [{pid,<0.30074.21>},{id,0},{mfargs,{brod_producer,start_link,[<0.30069.21>,<<\"brod-test-topic-1\">>,0,[]]}},{restart_type,{permanent,5}},{shutdown,5000},{child_type,worker}]","level":"error","error_logger":{"type":"supervisor_report","tag":"error_report"},"domain":["supervisor_report"]}
``` 

will brod work since nif is not supported?

**Reply by SergeTupchiy** [reply, marked as answer]
Hi @alexa22013, 
It's now looks similar to https://github.com/emqx/emqx/issues/10399 which I mentioned before, as crc32cer (brod dependency) also uses NIF (s function implemented in C). 
However, I cannot reproduce it now: I've just compiled a plugin with brod, added a basic producer sending a message to Kafka when a client connects to EMQX and it works fine (EMQX 5.1.0).

Could you please make sure that you compile the plugin on the same OS as used in EMQX environment where you try to upload it? 
For example if you compile the plugin on mac and than upload it to EMQX node powered by GNU/Linux, NIF can't work. 

As to the initial issue `"no such file or directory"`, I can confirm now that it's necessary to list all dependencies (including transitive ones) in `relx` config. Some of them may be omitted  and plugin will still work but only in case when all the omitted dependencies are present in EMQX (EMQX itself depends on them). That's why your plugin was able to start after specifying additional applications brod depends on. 
I have already registered a task to change it, so that it will be enough to only list direct dependencies, The ticket will be considered for development in future `emqx-plugin-template` releases.

**Reply by alexa22013** [reply]
Thank you @SergeTupchiy for your help. Followed the steps you mentiioned, I am able to build and run producer successfully. It will be helpful if its also documented in the plugin repo. Thanks again!

**Reply by SergeTupchiy** [reply]
Great, happy to help!
